aixcalibuha/sensitivity_analysis/sobol.py

```diff
@@ -68,13 +81,284 @@ def generate_samples(self):
+                        _conv_results.append(res_dict)
+                        tuples.append((local_class.name, goal, av))
+                    elif av in self.__analysis_variables_2:
+                        for tuner_para, res_dict in res_dict.items():
+                            _conv_results_2.append(res_dict)
+                            tuples_2.append((local_class.name, goal, av, tuner_para))
+        index = pd.MultiIndex.from_tuples(tuples=tuples,
+                                          names=['Class', 'Goal', 'Analysis variable'])
+        index_2 = pd.MultiIndex.from_tuples(tuples=tuples_2,
+                                            names=['Class', 'Goal', 'Analysis variable', 'Interaction'])
+        df = pd.DataFrame(_conv_results, index=index)
+        df_2 = pd.DataFrame(_conv_results_2, index=index_2)
+        result = (df, df_2)
+        return result
+
+    def _get_res_dict(self, result: dict, cal_class: CalibrationClass, analysis_variable: str):
         """
         Convert the result object to a dict with the key
         being the variable name and the value being the result
-        associated to self.analysis_variable.
+        associated to analysis_variable.
+        For second oder analysis variables the result is converted to a
+        dict with the key being the variable name and the value being another dict
+        with the variable names as the keys and the result associated to analysis_variable
+        from the interaction between the two variables.
         """
         names = self.create_problem(cal_class.tuner_paras)['names']
-        return {var_name: np.abs(res_val)
-                for var_name, res_val in zip(names,
-                                             result[self.analysis_variable])}
+        if analysis_variable in self.__analysis_variables_1:
+            res_dict_1 = {var_name: np.abs(res_val)
+                          for var_name, res_val in zip(names,
+                                                       result[analysis_variable])}
+            return res_dict_1
+        if analysis_variable in self.__analysis_variables_2:
+            result_av = result[analysis_variable]
+            for i in range(len(result_av)):
+                for j in range(len(result_av)):
+                    if i > j:
+                        result_av[i][j] = result_av[j][i]
+            res_dict_2 = {var_name: dict(zip(names, np.abs(res_val)))
+                          for var_name, res_val in zip(names,
+                                                       result_av)}
+            return res_dict_2
+
+    @staticmethod
+    def plot_second_order(result: pd.DataFrame, **kwargs):
+        """
+        Plot sensitivity results of second order analysis variables.
+        For each calibration class and goal one figure of a 3d plot is created
+        with the barplots of the interactions for each parameter.
+        Only working for more than 2 parameter.
+
+        :param pd.DataFrame result:
```

Comment:
Is sobol always the only method in need for all this plotting? If it could be used in PAWN, FAST or morris, shouldn't this be in some general place? And: Why is it a class-method if public and static? A normal method would do it as well, or?

Reply:
The plot function in the SobolAnalyzer can always only be used with the second order results of this method. I would like to keep them as class-methods because in the most cases they would be used after the SobolAnalyzer was used. Secondly, they are so specific to the results of this class that it is more clear this way that they can only be used with results of this class.